# 오류 코드를 예외로 변경

```jsx
if (data)
	return ShippingRules(data);
else
	return -23;
```

↓

```jsx
if (data)
	return ShippingRules(data);
else
	throw new OrderProcessingError(-23);
```

<aside>
⚠️ *"... 예외는 정교한 메커니즘이지만 대다수의 다른 정교한 메커니즘과 같이 정확하게 사용할 때만 최고의 효과를 낸다. ... 프로그램의 정상 동작 범주에 들지 않는 오류를 나타낼 때만 쓰여야 한다. ... 예외를 던지는 코드를 프로 그램 종료 코드로 바꿔도 여전히 정상 동작할지를 따져보는 것... 정상 동작하지 않을 것 같다면 예외를 사용하지 말라는 신호이다. ..."*

</aside>

### 절차

1. 콜스택 상위에 해당 예외를 처리할 예외 핸들러를 작성
    - 이 핸들러는 처음엔 모든 예외를 다시 던지게 작성
    - 적절한 처리를 해주는 핸들러 존재시 지금의 콜스택도 처리 가능하게 확장
    - 예시

    ```jsx
    //// 최상위
    **const status = calculateShippingCosts(orderData);
    if (status < 0) errorList.push({order: orderData, errorCode: status});**

    function calculateShippingCosts(anOrder) {
    	const shippingRules = localShippingRules(anOrder.country);
    	if (shippingRules < 0) return shippingRules; // 오류 전파
    	...
    }

    function localShippingRules(country) {
    	const data = countryData.ShippingRules[country];
    	if (data) return ShippingRules(data);
    	else return -23;
    }
    ```

    ↓

    ```jsx
    //// 최상위
    **let status;
    try{
    	status = calculateShippingCosts(orderData);
    } catch(e){
    	throw e;
    }**
    if (status < 0) errorList.push({order: orderData, errorCode: status});
    ...
    ```

2. 테스트
3. 해당 오류 코드를 대체할 예외와 그 밖의 예외를 구분할 식별 방법을 탐색
    - 예시

    ```jsx
    class OrderProcessingError extends Error { //dart 같은 경우는 'extends Exception'
    		constructor(errorCode) {
    		super('주문 처리 오류 $errorCode');
    		this.code = errorCode;
    	}
    	get name() => "OrderProcessingError";
    }
    ```

4. 정적 검사
5. catch절을 수정해 직접 처리 가능한 예외는 적절히 대처 OR 그렇지 않은 예외는 다시 던진다.
    - 예시

    ```jsx
    //// 최상위
    let status;
    try{
    	status = calculateShippingCosts(orderData);
    } catch(e){
    	**if(e instanceof OrderProcessingError)
    		errorList.push({order: orderData, errorCode: e.code});
    	throw e;**
    }
    if (status < 0) errorList.push({order: orderData, errorCode: status});
    ...
    ```

6. 테스트
7. 오류 코드를 반환하는 곳 모두에서 예외를 던지도록 수정
    - (1수정 1테스트)
    - 예시

    ```jsx
    function localShippingRules(country) {
    	const data = countryData.ShippingRules[country];
    	if (data) return ShippingRules(data);
    	else **throw new OrderProcessingError(-23);**
    }
    ```

8. 모두 수정 했다면 그 오류 코드를 콜스택 위로 전달하는 코드를 모두 제거한다.
    - (1수정 1테스트)
    - 예시
        1. 오류 코드 검사하는 부분을 함정으로 교체 ↓
        2. 함정에 걸렸는지 테스트후 제거 할 것

    ```jsx
    function calculateShippingCosts(anOrder) {
    	const shippingRules = localShippingRules(anOrder.country);
    	if (shippingRules < 0) **throw new Error("오류코드가 아직 남아있습니다.");**
    	...
    }
    ```
